src/Razor/src/Microsoft.VisualStudio.LanguageServer.ContainedLanguage/VirtualDocumentFactoryBase.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServer.ContainedLanguage
+{
+    internal abstract class VirtualDocumentFactoryBase : VirtualDocumentFactory
+    {
+        private readonly ITextBufferFactoryService _textBufferFactory;
+        private readonly ITextDocumentFactoryService _textDocumentFactory;
+        private readonly FileUriProvider _fileUriProvider;
+
+        protected IContentTypeRegistryService ContentTypeRegistry { get; }
+
+        public VirtualDocumentFactoryBase(
+            IContentTypeRegistryService contentTypeRegistry,
+            ITextBufferFactoryService textBufferFactory,
+            ITextDocumentFactoryService textDocumentFactory,
+            FileUriProvider filePathProvider)
+        {
+            if (contentTypeRegistry is null)
+            {
+                throw new ArgumentNullException(nameof(contentTypeRegistry));
+            }
+
+            if (textBufferFactory is null)
+            {
+                throw new ArgumentNullException(nameof(textBufferFactory));
+            }
+
+            if (textDocumentFactory is null)
+            {
+                throw new ArgumentNullException(nameof(textDocumentFactory));
+            }
+
+            if (filePathProvider is null)
+            {
+                throw new ArgumentNullException(nameof(filePathProvider));
+            }
+
+            ContentTypeRegistry = contentTypeRegistry;
+
+            _textBufferFactory = textBufferFactory;
+            _textDocumentFactory = textDocumentFactory;
+            _fileUriProvider = filePathProvider;
+        }
+
+        protected abstract IContentType LanguageLSPContentType { get; }
```

Comment:
```suggestion
        protected abstract IContentType LanguageContentType { get; }
```

Reply:
I kind of like the distinction here between "normal" and "lsp", but I don't feel that strongly about it. I'll go ahead with your suggestion.

---
In reply to: [452550033](https://github.com/dotnet/aspnetcore-tooling/pull/2201#discussion_r452550033) [](ancestors = 452550033)